### PR TITLE
Laboratory work #3, Shhanukova Milana - 19FPL1

### PR DESCRIPTION
**Pull request by MilanaShhanukova**

### CODE REVIEW

**Comment by WhiteJaeger** [comment]
Please rename using the template so we can see whose PR this is.

File: lab_2/main.py [hunk 1]
```diff
@@ -1,18 +1,42 @@
 """
 Longest common subsequence problem
 """
+from tokenizer import tokenize
 
 
-def tokenize_by_lines(text: str) -> tuple:
```
Comment (WhiteJaeger) [inline lab_2/main.py:6]: This function is needed in tests and for working with big files.

File: lab_2/main.py [hunk 2]
```diff
@@ -1,18 +1,42 @@
...
-    """
-    Splits a text into sentences, sentences – into tokens,
-    converts the tokens into lowercase, removes punctuation
-    :param text: the initial text
-    :return: a list of sentences with lowercase tokens without punctuation
-    e.g. text = 'I have a cat.\nHis name is Bruno'
-    --> (('i', 'have', 'a', 'cat'), ('his', 'name', 'is', 'bruno'))
-    """
-    pass
+def checker(object_to_check, expected_type):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: Functions usually have verb-like names:
Like `check_object_type`

File: lab_2/main.py [hunk 3]
```diff
@@ -68,7 +150,21 @@ def calculate_plagiarism_score(lcs_length: int, suspicious_sentence_tokens: tupl
...
-    pass
+
+    conditions = [checker(lcs_length, int),
+                  checker(suspicious_sentence_tokens, tuple)]
+    if not all(conditions):
+        return -1
+    if lcs_length != 0 and len(suspicious_sentence_tokens) == 0:
+        return 0.0
+    if lcs_length == 0 and len(suspicious_sentence_tokens) == 0:
+        return -1
```
Comment (WhiteJaeger) [inline lab_2/main.py]: These checks can be shortened, like `not object` or `not len(object)`.

File: lab_2/main.py [hunk 4]
```diff
@@ -68,7 +150,21 @@ def calculate_plagiarism_score(lcs_length: int, suspicious_sentence_tokens: tupl
...
+    if not(lcs_length <= len(suspicious_sentence_tokens)):
+        return -1
+    if lcs_length == 0:
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `lcs_length == 0` can be shortened.

File: lab_2/main.py [hunk 5]
```diff
@@ -81,7 +177,35 @@ def calculate_text_plagiarism_score(original_text_tokens: tuple, suspicious_text
...
+        changed_sent_tokens = list(original_text_tokens)
+        for i in range(len(suspicious_text_tokens) - len(original_text_tokens)):
+            changed_sent_tokens.append(())
+        original_text_tokens = tuple(changed_sent_tokens)
+    elif len(suspicious_text_tokens) < len(original_text_tokens):
+        original_text_tokens = original_text_tokens[:len(suspicious_text_tokens)]
+
+    lcs = 0
+    # counting
+    for l in range(len(original_text_tokens)):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `l` is not very descriptive name, right?

File: lab_2/main.py [hunk 6]
```diff
@@ -92,7 +216,36 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
     :return: a tuple with tuples of indexes
     """
-    pass
+    conditions = [checker(original_sentence_tokens, tuple),
+                  checker(suspicious_sentence_tokens, tuple),
+                  checker(lcs, tuple)]
+    if not(all(conditions)):
+        return ()
+
+    def count_difference(tokens):
```
Comment (WhiteJaeger) [inline lab_2/main.py:246]: It is better not to nest functions. I suggest moving it outside and using it like utility function here.

File: lab_2/main.py [hunk 7]
```diff
@@ -119,26 +289,47 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
 
+    lcs = []
+    # conditions to find the subsequence
+    conditions = [
+                  checker(first_sentence_tokens, tuple),
+                  checker(second_sentence_tokens, tuple)]
+    if not all(conditions):
+        return ()
 
-def tokenize_big_file(path_to_file: str) -> tuple:
```
Comment (WhiteJaeger) [inline lab_2/main.py:138]: Please, leave function descriptions as they are.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_2/find_lenght_test.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
+# pylint: skip-file
```
Comment (WhiteJaeger) [inline lab_2/find_lenght_test.py:1]: What is this file?

File: lab_2/main.py [hunk 9]
```diff
@@ -12,7 +13,45 @@ def tokenize_by_lines(text: str) -> tuple:
     e.g. text = 'I have a cat.\nHis name is Bruno'
     --> (('i', 'have', 'a', 'cat'), ('his', 'name', 'is', 'bruno'))
     """
-    pass
+    text = text.split("\n")
+    tokenized_text = [tuple(tokenize(sent)) for sent in text]
+    tokenized_text = [sent for sent in tokenized_text if sent != ()]
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `if sent != ()` can be shortened.

File: lab_2/main.py [hunk 10]
```diff
@@ -119,26 +304,61 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
-    Finds a length of the longest common subsequence using an optimized algorithm
-    When a length is less than the threshold, it becomes 0
-    :param first_sentence_tokens: a tuple of tokens
-    :param second_sentence_tokens: a tuple of tokens
-    :param plagiarism_threshold: a threshold
+    Finds a length of the longest common subsequence using the Hirschberg's algorithm
+    At the same time, if the first and last tokens coincide,
+    they are immediately added to lcs and not analyzed
+    :param first_sentence_tokens: a list of tokens
+    :param second_sentence_tokens: a list of tokens
```
Comment (WhiteJaeger) [inline lab_2/main.py]: Please, leave function descriptions as they are.

File: lab_2/main.py [hunk 11]
```diff
@@ -119,26 +304,61 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
 
 
 def tokenize_big_file(path_to_file: str) -> tuple:
-    """
-    Reads, tokenizes and transforms a big file into a numeric form
-    :param path_to_file: a path
-    :return: a tuple with ids
-    """
-    pass
+    with open("data.txt") as first_file, open("data_2.txt") as second_file:
```
Comment (WhiteJaeger) [inline lab_2/main.py]: You do need to open these particular files. 
You should work with `path_to_file` argument.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_2/main.py [hunk 12]
```diff
@@ -119,26 +312,62 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
 def tokenize_big_file(path_to_file: str) -> tuple:
-    """
-    Reads, tokenizes and transforms a big file into a numeric form
-    :param path_to_file: a path
-    :return: a tuple with ids
-    """
-    pass
+    with open(path_to_file, encoding="utf-8") as first_file, open(path_to_file, encoding="utf-8") as second_file:
+        first_lines = (tokenize_by_lines(line_1) for line_1 in first_file)
+        second_lines = (tokenize_by_lines(line_2) for line_2 in second_file)
```
Comment (WhiteJaeger) [inline lab_2/main.py]: Why do you open the same file twice?

File: lab_2/main.py [hunk 13]
```diff
@@ -119,26 +312,62 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
+
+        lengths = (find_lcs_length_optimized(first_line, second_line)
+                   for first_line, second_line in zip(first_lines, second_lines))
```
Comment (WhiteJaeger) [inline lab_2/main.py]: These would be identical files, right?

File: lab_2/main.py (hunk 12 above)
Comment (MilanaShhanukova) [reply lab_2/main.py]: i've changed this already

File: lab_2/main.py (hunk 13 above)
Comment (MilanaShhanukova) [reply lab_2/main.py]: i've changed this already

**Pull request closed without merging**